### PR TITLE
problem solving elaboration

### PR DESCRIPTION
**Pull request by anharrington**
First commit is just cleaning up WithStatements.
Main one is a major rewrite of the start of the debugging chapter.  Seemed like my stuff fit with the title How to be a Successful Programmer, and I liked the location in the book just after basic data.  I kept some of what was in the section before, but dumped some duplication with the next section, and consciously dumped the part about 99% of programming is debugging - really suggests minimizing understanding and planning.

### CODE REVIEW

File: _sources/Debugging/intro-HowtobeaSuccessfulProgrammer.rst [hunk 1]
```diff
@@ -7,10 +7,230 @@
...
 
+.. index:: problem solving strategy
+
+.. _learn-solve:
+
 How to be a Successful Programmer
-===============================================
+==================================
+
+This section might have been placed earlier,
```
Comment (bnmnetp) [inline _sources/Debugging/intro-HowtobeaSuccessfulProgrammer.rst]: I couldn't imagine putting this earlier.

File: _sources/Debugging/intro-HowtobeaSuccessfulProgrammer.rst [hunk 2]
```diff
@@ -7,10 +7,230 @@
...
+It can seem overwhelming.  It need not be. Take a breath.  
+
+First basic language syntax:  When learning any new language, there is a lot to take in.  
+We introduce Python in chunks.  For a while there will always be the new current syntax topic coming.  
+You do NOT need to memorize *everything* immediately!  
+
+- Some things that you use rarely, you may never memorize, like,
+  "Is ``exec`` a keyword?"  
+  At *some* point that might be useful.
+  Can you find it?  It happens to be in :ref:`variables-and-heywords`.
```
Comment (bnmnetp) [inline _sources/Debugging/intro-HowtobeaSuccessfulProgrammer.rst]: variables-and-keywords I think

File: _sources/Debugging/intro-HowtobeaSuccessfulProgrammer.rst [hunk 3]
```diff
@@ -7,10 +7,230 @@
...
 
-Programming is an odd thing in a way.  Here is why.  As programmers we spend 99% of our time trying to get our program to work.  We struggle, we stress, we spend hours deep in frustration trying to get our program to execute correctly.  Then when we do get it going we celebrate, hand it in, and move on to the next homework assignment or programming task.  But here is the secret, when you are successful, you are happy, your brain releases a bit of chemical that makes you feel good.  You need to organize your programming so that you have lots of little successess.  It turns out your brain doesn't care all that much if you have successfully written hello world, or a fast fourier transform (trust me its hard) you still get that little release that makes you happy.  When you are happy you want to go on and solve the next little problem.  Essentially I'm telling you once again, start small, get something small working, and then add to it.
```
Comment (bnmnetp) [inline _sources/Debugging/intro-HowtobeaSuccessfulProgrammer.rst]: I like this paragraph.  I would rather keep it.

Comment (anharrington) [reply _sources/Debugging/intro-HowtobeaSuccessfulProgrammer.rst]: This was originally in 3.1.  Part of this is repeated.  3.2 has

Every time you have a little success, your brain releases a tiny bit of chemical that makes you happy. So, you can keep yourself happy and make programming more enjoyable by creating lots of small victories for yourself.

In latest push I replaced this duplicated part by your whole paragraph, now in 3.2.

Comment (anharrington) [reply _sources/Debugging/intro-HowtobeaSuccessfulProgrammer.rst]: I think pushing it was enough to get it added to this still unaccepted pull request.

**Comment by bnmnetp** [comment]
Closing this as most of this has been merged through other PRs.

**Pull request closed without merging**